### PR TITLE
[ISSUE #555] when no msg or offset not match correct local offset.

### LINKED ISSUE #555: push_consumer not implement correctTagsOffset method
hi 
1. [push_consumer](https://github.com/apache/rocketmq-client-go/blob/master/consumer/push_consumer.go#L717) 没有实现correctTagsOffset方法导致在broker没有新消息时，通知客户端修改offset，但是客户端却未能更新。
参考java[代码](https://github.com/apache/rocketmq/blob/master/client/src/main/java/org/apache/rocketmq/client/impl/consumer/DefaultMQPushConsumerImpl.java#L481)如下：
```
private void correctTagsOffset(final PullRequest pullRequest) {
    if (0L == pullRequest.getProcessQueue().getMsgCount().get()) {
        this.offsetStore.updateOffset(pullRequest.getMessageQueue(), pullRequest.getNextOffset(), true);
    }
}
```

2. 另外[case primitive.PullNoNewMsg](https://github.com/apache/rocketmq-client-go/blob/master/consumer/push_consumer.go#L692)这种情况也需要调用correctTagsOffset方法。

参考java的实现：
image

请及时修复，谢谢

### PR DESCRIPTION
**Pull request by gaoyf**
ref #555

### CODE REVIEW

**Review by ShannonDing — APPROVED** [review]

File: consumer/push_consumer.go [hunk 1]
```diff
@@ -692,6 +692,7 @@ func (pc *pushConsumer) pullMessage(request *PullRequest) {
 		case primitive.PullNoNewMsg:
 			rlog.Debug(fmt.Sprintf("Topic: %s, QueueId: %d no more msg, current offset: %d, next offset: %d",
```
Comment (xujianhai666) [inline consumer/push_consumer.go:693]: I found that u should correct next offset
```
                        case NO_NEW_MSG:
                            pullRequest.setNextOffset(pullResult.getNextBeginOffset());
                            DefaultMQPushConsumerImpl.this.correctTagsOffset(pullRequest);
                            DefaultMQPushConsumerImpl.this.executePullRequestImmediately(pullRequest);
```

**Comment by ingtube-zhangkun** [comment]
consumingMsgOrderlyTreeMap  may be not empty !

**Review by qianlongzt — APPROVED** [review]

**Comment by qianlongzt** [comment]
可以合并吗？

File: consumer/push_consumer.go [hunk 2]
```diff
@@ -715,7 +717,12 @@ func (pc *pushConsumer) pullMessage(request *PullRequest) {
 }
 
 func (pc *pushConsumer) correctTagsOffset(pr *PullRequest) {
-	// TODO
+	if pr.pq.cachedMsgCount == 0 {
```
Comment (qianlongzt) [inline consumer/push_consumer.go:720]:
```suggestion
	if atomic.LoadInt64(&pr.pq.cachedMsgCount) == 0 {
```

File: consumer/push_consumer.go [hunk 3]
```diff
@@ -715,7 +717,12 @@ func (pc *pushConsumer) pullMessage(request *PullRequest) {
 }
 
 func (pc *pushConsumer) correctTagsOffset(pr *PullRequest) {
-	// TODO
+	if atomic.LoadInt64(&pr.pq.cachedMsgCount) == 0 {
```
Comment (qianlongzt) [inline consumer/push_consumer.go:720]: check more

```suggestion
	if atomic.LoadInt64(&pr.pq.cachedMsgCount) == 0 {
		if pr.pq.IsDroppd() {
			return
		}
		if !pr.pq.IsLock() {
			return
		}
		if pr.pq.isLockExpired() {
			return
		}
```

**Comment by qianlongzt** [comment]
cachedMsgCount should alway use by atomic


https://github.com/ingtube/rocketmq-client-go/commit/1864f166a2fa82c02816accf04d175094764c3c3

**Pull request closed without merging**